### PR TITLE
dyno: make params without values be considered unknown

### PR DESCRIPTION
**Pull request by DanilaFe**
This is needed for the resolution of if-expressions with `param`
conditions. An if-expression depending on a `param bool` without a value
should be considered unknown (as is the case for an if-expression
depending on any other unknown value).

This commit also implements special-case behavior for `param _tuple`
types, which have their tuple elements' values encoded in the
`QualifiedType`s of the `TupleType`, but not stored into a `Param*`.

There is discussion in https://github.com/chapel-lang/chapel/issues/20487 
about how to properly represent param tuples. It is likely that in the
event that we add param tuples to Dyno, we will settle on a representation
that includes a special `Param` that contains tuple values. This would
preclude the need for special-case handling in isUnknown. In the meantime,
though, we don't have Param tuples (and they're not in the production compiler),
so this workaround seems appropriate. 

Reviewed by @benharsh  - thanks!

TESTING:
- [x] Paratest. some failures on this branch are fixed on main, but nothing
   otherwise.

Signed-off-by: Danila Fedorin <daniel.fedorin@hpe.com>

### CODE REVIEW

File: compiler/dyno/lib/types/QualifiedType.cpp [hunk 1]
```diff
@@ -50,6 +51,22 @@ bool QualifiedType::isParamFalse() const {
   return false;
 }
 
+bool QualifiedType::isParamKnownTuple() const {
+  if (kind_ == Kind::PARAM && type_ != nullptr) {
+    if (auto tt = type_->toTupleType()) {
+      // Check if all type actuals are known params
+      for (int i = 0; i < tt->numElements(); i++) {
```
Comment (benharsh) [inline compiler/dyno/lib/types/QualifiedType.cpp:58]: I think this is fine for now, but in the future we should compute this once in the TupleType constructor and cache the result.

**Review by benharsh — APPROVED** [review]
Looks good to me! 👍

File: compiler/dyno/lib/types/QualifiedType.cpp (hunk 1 above)
Comment (DanilaFe) [reply compiler/dyno/lib/types/QualifiedType.cpp]: I figured it was easy enough to fix 😄

**Review by benharsh — APPROVED** [review]

**Comment by bradcray** [comment]
@DanilaFe : What does "without values" mean in this context?

**Comment by DanilaFe** [comment]
Something like the following: 
```Chapel
proc f(param a: bool, b: if a then int else string) {}
```
The type of `a` is `param bool`, but it's missing the `=.. ` part. That's the "no value" I'm talking about here.

**Comment by bradcray** [comment]
OK, so like "param arguments that don't have a default value" essentially (vs. cases like standalone param variables like `param x: int;`)

**Pull request merged**